packages/serverpod_test/lib/src/test_serverpod.dart

```diff
@@ -112,22 +124,23 @@ class TestServerpod<T extends InternalTestEndpoints> {
     // Ignore output from the Serverpod constructor to avoid spamming the console.
     // Should be changed when a proper logger is implemented.
     // Tracked in issue: https://github.com/serverpod/serverpod/issues/2847
-    IOOverrides.runZoned(
-      () {
-        _serverpod = Serverpod(
-          _getServerpodStartUpArgs(
-            runMode: runMode,
-            applyMigrations: applyMigrations,
-            loggingMode: serverpodLoggingMode,
-          ),
-          serializationManager,
-          endpoints,
+    _buildServerpodAndInitializeEndpoints = () => IOOverrides.runZoned(
+          () {
+            var serverpod = Serverpod(
+              _getServerpodStartUpArgs(
+                runMode: runMode,
+                applyMigrations: applyMigrations,
+                loggingMode: serverpodLoggingMode,
+              ),
+              serializationManager,
+              endpoints,
+            );
+            endpoints.initializeEndpoints(serverpod.server);
+            return serverpod;
+          },
+          stdout: () => NullStdOut(),
+          stderr: () => NullStdOut(),
         );
-      },
-      stdout: () => NullStdOut(),
-      stderr: () => NullStdOut(),
-    );
-    endpoints.initializeEndpoints(_serverpod.server);
     testEndpoints.initialize(serializationManager, endpoints);
```

Comment:
Question: Just to make sure we didn't miss something, is it fine iof endpoints are passed here to the initialize method even if we have not initialized them?

Reply:
The endpoints only needs to be initialized before being called. Since we will always build out a session before calling an endpoint, which will trigger this lazy build, it can never happen. 

As a side note, this case is if the user does an un-awaited call to an endpoint before the setups/tests which _may_ enter the code path just described depending on execution order.